library/psa_crypto.c

```diff
@@ -1106,6 +1106,17 @@ psa_status_t psa_destroy_key(mbedtls_svc_key_id_t key)
      * fully destroyed. */
     PSA_THREADING_CHK_GOTO_EXIT(mbedtls_mutex_lock(
                                     &mbedtls_threading_key_slot_mutex));
+
+    if (slot->state == PSA_SLOT_PENDING_DELETION) {
+        /* Another thread has destroyed the key between us locking the slot
+         * and us gaining the mutex. Unregister from the slot,
+         * and report that the key does not exist. */
+        status = psa_unregister_read(slot);
+
+        PSA_THREADING_CHK_RET(mbedtls_mutex_unlock(
+                                  &mbedtls_threading_key_slot_mutex));
+        return (status == PSA_SUCCESS) ? PSA_ERROR_INVALID_HANDLE : status;
+    }
```

Comment:
Any way of adding a test for this?

Reply:
Not entirely sure if I am interpreting this correctly, but the tests that found this issue will come in a seperate PR this week.

Reply:
That is the plan. The test which catches this also tests concurrently using the same key in a myriad of ways. Doing this requires some overhaul to `psa_exercise_key.c` which I am currently working on. So my current plan is to put that in a different PR

Reply:
Ah, okay, thanks for the reassurance